Async memcpy for DtoD

tomstitt opened the pull request:
Use `HipMemcpyDtoDAsync` (on the default stream) instead of `HipMemcpyDtoD` to avoid unnecessary host blocking. I noticed this in the profiler and after reading some documentation it seems like cuda and hip have different behavior for synchronous DtoD memcpys.

The cuda docs specifically say that "for transfers from device memory to device memory, no host-side synchronization is performed," for synchronous memcpy ([see the cuda docs](https://docs.nvidia.com/cuda/cuda-runtime-api/api-sync-behavior.html#api-sync-behavior__memcpy-sync)), while for hip every synchronous memcpy will:
- wait for all previous commands to complete.
- will not return control back to host until the command completes.
- ([see the rocm/hip docs](http://rocm-developer-tools.github.io/HIP/Synchonization.html))

I've only tested this for the Umpire device memory space and we see a performance improvement resulting from better kernel packing in a iterative routine with a few kernels followed by a DtoD transfer followed by a few kernels followed by a DtoD transfer followed by ... .  

From what I understand this change is okay because the default stream waits for "tasks" on other streams to complete before running so this will be synchronous wrt other device tasks on non-default streams (if we ever add support for that in mfem). Hard to find this in the docs but someone from amd mentioned it [here](https://github.com/ROCm-Developer-Tools/HIP/issues/129).

I am happy to make this change for other hip device memory spaces if desired.
<!--GHEX{"id":2780,"author":"tomstitt","editor":"tzanio","reviewers":["artv3","camierjs","v-dobrev"],"assignment":"2022-01-21T15:37:35-08:00","approval":"2022-01-27T23:28:39.337Z","merge":"2022-01-31T01:17:59.189Z"}XEHG--><!--GHEXTABLE-->
 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#2780](https://github.com/mfem/mfem/pull/2780) | @tomstitt | @tzanio | @artv3 + @camierjs + @v-dobrev | 01/21/22 | 01/27/22 | 01/30/22 | |
<!--ELBATXEHG-->

camierjs commented on general/mem_manager.cpp:
```diff
@@ -593,7 +593,7 @@ class UmpireDeviceMemorySpace : public DeviceMemorySpace,
 #endif
 #ifdef MFEM_USE_HIP
-      return HipMemcpyDtoD(dst, src, bytes);
+      return HipMemcpyDtoDAsync(dst, src, bytes);
```
I didn't realized this was only for the `UmpireDeviceMemorySpace` when we discussed this earlier.
Do you know what would trigger the `rm.copy(dst, const_cast<void*>(src), bytes)` call?
If `Umpire` has the location knowledge, maybe it would be more appropriate to let it handle the copy.

tomstitt replied:
> I didn't realized this was only for the `UmpireDeviceMemorySpace` when we discussed this earlier.

Yea, it seemed to much of a change to do everything. I would be awesome if we could just extend the `UmpireDeviceMemorySpace` locally ourselves and do this but that's a whole other set of changes

> Do you know what would trigger the `rm.copy(dst, const_cast<void*>(src), bytes)` call?

If I'm understanding your question it looks like it is commented out right now so it would never trigger and that we used to call it but it was changed here: https://github.com/mfem/mfem/commit/c77720a203b43e944a555eff821391d7feea5421 . But maybe you are asking what Umpire _would_ do and I'm not sure of that but I will check

tomstitt replied:
updated the `HipDeviceMemorySpace` as well

v-dobrev commented:
I think we should to use the async D2D copy in class `HipDeviceMemorySpace` too.

Also, it will be nice to have a short comment at those calls to explain why we do that (and maybe mention this PR in the comment too.)

tomstitt commented:
> I think we should use the async D2D copy in class `HipDeviceMemorySpace` too.

I added it there too

> Also, it will be nice to have a short comment at those calls to explain why we do that (and maybe mention this PR in the comment too.)

I added a short comment in both spots and also linked this PR

tzanio commented:
Merged in `next` for testing...

tzanio reviewed (approved):
Thanks @tomstitt !
